api/lib/opentelemetry/trace/samplers/always_sample_sampler.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'opentelemetry/trace/samplers/basic_decision'
+
+module OpenTelemetry
+  module Trace
+    module Samplers
+      # The {AlwaysSampleSampler} always returns a true sampling decision.
+      class AlwaysSampleSampler
+        ALWAYS_SAMPLE_DECISION = BasicDecision.new(decision: true)
+
+        # Returns the sampling {Decision} for a {Span} to be created
+        #
+        # @param [SpanContext] span_context The {SpanContext} of a parent span,
+        #   typically extracted from the wire. Can be nil.
+        # @param [Boolean] extracted_context True if span_context was extracted
+        #   from the wire. Can be nil.
+        # @param [Integer] trace_id The trace_id of the {Span} to be created
+        # @param [Integer] span_id The span_id of the {Span} to be created
+        # @param [String] span_name Name of the {Span} to be created
+        # @param [Array<Links>] links A collection of links to be associated
+        #   with the {Span} to be created
+        # @return [Decision] The sampling decision
+        def should_sample(span_context: nil,
+                          extracted_context: nil,
+                          trace_id:,
+                          span_id:,
+                          span_name:,
+                          links:)
+          ALWAYS_SAMPLE_DECISION
```

Comment:
Should we type-check the arguments here? On one hand, this might well be used in production, and we shouldn't add overhead there. On the other hand, we want to be able to catch errors in use.

Reply:
We can. In that case I would make a Sampler abstract class with template methods and the type checking logic.

Reply:
I implemented type checking in the abstract `Sampler` class. It's a bit cumbersome and did not go over especially with RuboCop. I'm open to suggestions here.

Reply:
I don't think it'd help with RuboCop, but would a simple `super` be sufficient here, with type checking implemented in `Sampler#should_sample`?

Reply:
We could do that. Would you have any qualms about `Sampler#should_sample` only doing type checking and not having a `Decision` return value?

Reply:
No qualms. Returning `nil` from `Sampler#should_sample` seems OK, even if we document it to return a `Decision`. An alternative would be a type-checking mixin, but I'm not sure that's any better.